Added clear_resolver method to OmegaConf

sugatoray opened the pull request:
This closes #769 .

- [x] Added `OmegaConf.clear_resolver()` method to `omegaconf.py`.  :green_circle: 
  - This allows the user to remove a single resolver.
- [x] Added tests to `tests/test_omegaconf.py::test_clear_resolver`. :green_circle: 
- [x] Added documentation for `OmegaConf.clear_resolvers` and `OmegaConf.clear_resolver`. :green_circle:

sugatoray commented:
Hi @omry,

I hope this functionality will give people ability to selectively remove certain custom resolvers, after adding them. Please let me know your thoughts on it. Thank you.

Jasha10 commented:
Thanks for submitting this @sugatoray :)
I'll take a look at it shortly.

sugatoray commented:
@Jasha10 If you see my current implementation, you will find that the constant list, `DEFAULT_RESOLVER_NAMES` is being manually created. This means in future, if you were to add/remove a default resolver (like removing deprecated resolver `env`), you will have to manually update this list. 

:question: What if I move it inside the method that defines the default resolvers and dynamically create `DEFAULT_RESOLVER_NAMES` as:

```
DEFAULT_RESOLVER_NAMES = list(BaseConfig._resolvers.keys())
```

This will ensure that this list is always populated with the default resolver names and no manual maintenance of it will be necessary either.

Jasha10 commented:
> question What if I move it inside the method that defines the default resolvers...

Yes, that could work. That being said, please see [Omry's suggestion](https://github.com/omry/omegaconf/issues/769#issuecomment-884302355) in the linked issue; we might be able to skip the `DEFAULT_RESOLVER_NAMES` variable all together.

sugatoray commented:
@Jasha10 I just pushed the changes to the PR. Please review it and let me know if any changes should be made. Thank you.

Jasha10 commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,93 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+            # expected
+            dict(pre_addition=True, post_addition=True, post_removal=False),
+        ),
+        # tuple
+        (
+            # cfg_params: Check with a nonexisting resolver
+            dict(),
+            # clear_resolver_params
+            dict(name="abcdef"),
+            # expected
+            dict(pre_addition=False, post_addition=False, post_removal=False),
+        ),
```
To be consistent with the other tests used in this codebase, it would be good to use `param(..., id="test_name")` for each test case, e.g. see the `test_resolve` function above.
For example, the `id` for this test case can be `"nonexisting_resolver"`.

In general, please try to make the code style consistent with the other tests in this repository.

sugatoray replied:
Will do.

sugatoray replied:
Done!

Jasha10 commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,93 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+            # cfg_params
+            dict(
+                name="TR1",
+                resolver=lambda x: str(x).lower(),
+                use_cache=False,
+                replace=True,
+            ),
+            # clear_resolver_params
+            dict(name="TR1"),
+            # expected
+            dict(pre_addition=False, post_addition=True, post_removal=False),
+        ),
+        # tuple
+        (
+            # cfg_params
+            dict(
+                name="TR2",
+                resolver=lambda x: str(x).upper(),
+                use_cache=False,
+                replace=True,
+            ),
+            # clear_resolver_params
+            dict(name="TR2"),
+            # expected
+            dict(pre_addition=False, post_addition=True, post_removal=False),
+        ),
+        # tuple
+        (
+            # cfg_params
+            dict(
+                name="TR3",
+                resolver=lambda x: str(x).upper(),
+                use_cache=False,
+                replace=False,
+            ),
+            # clear_resolver_params
+            dict(name="TR3"),
+            # expected
+            dict(pre_addition=False, post_addition=True, post_removal=False),
+        ),
```
I don't think we need to test all of these combinations of `use_cache`/`resolver`/`replace` -- we are trying to test `OmegaConf.clear_resolver`, not `OmegaConf.register_resolver`. It should be good enough to test these cases:
  - a custom resolver (registered by the user)
  - a default resolver (e.g. `oc.env`)
  - a non-existing resolver

sugatoray replied:
Ok. I will slim it down. Thank you.

sugatoray replied:
Done!

Jasha10 commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,93 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+
+    assert OmegaConf.clear_resolver(**clear_resolver_params)
+
+    assert expected["post_removal"] == OmegaConf.has_resolver(name)
```
In all of the test cases, `expected["post_removal"]` is `False`. I think we should just use `assert not OmegaConf.has_resolver(name)` here.

sugatoray replied:
I had a sense perhaps some of the test cases will need change, as first I wanted to make sure the implementation of clear_resolver seems okay. Let me slim the tests down and then I think it will end up something like you suggested.

sugatoray replied:
Done!

Jasha10 reviewed (changes requested):
Thanks @sugatoray. A few comments:

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,93 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+        OmegaConf.register_new_resolver(**cfg_params)
+        assert expected["post_addition"] == OmegaConf.has_resolver(name)
+
+    assert OmegaConf.clear_resolver(**clear_resolver_params)
```
Okay. Makes sense. However, I thought @omry suggested the bool output for True when resolver is removed (*if something has been removed, then it does not exist anymore; and it’s current state is the same as something that never existed*) and False when not removed for whatever reason.

Jasha10 replied:
I guess my point is that it is not very useful to return a bool 
> output of `True` when resolver is removed

Agreed.

 > or for when the resolver did not exist in the first place

I think it would be more useful to output `False` if the resolver did not exist in the first place.
The reason is that returning `True` in all cases (which is what the function does currently) is not very useful.

Jasha10 replied:
From Omry's [original comment](https://github.com/omry/omegaconf/issues/769#issuecomment-884302355) about returning a bool:

> A simple API for clear resolver can be:
> 
> ```python
> def clear_resolver(name: str) -> bool:
>   ...
> ```
> 
> This will remove a resolver and return False if it's not found, allowing the user to handle this case.
> It will allow a user to remove all resolvers, including built in ones. I think this is okay.

In particular, "return False if it's not found"

sugatoray replied:
@Jasha10 Thank you, for pointing it out. I shouldn't have missed it. I will make the change.

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,93 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+    cfg_params: Any, clear_resolver_params: Any, expected: Any
+) -> None:
+
+    OmegaConf.clear_resolvers()
```
Thank you, will make this change.

sugatoray replied:
Done!

Jasha10 commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,66 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+    ],
+)
+def test_clear_resolver(
+    restore_resolvers: Any, cfg_params: Any, clear_resolver_params: Any, expected: Any
```
Since `clear_resolver_params` is a dict whose only key is `"name"`, it would be simpler to just pass in the string as an argument to the test function:

```python
def test_clear_resolver(restore_resolvers: Any, cfg_params: Any, name: str, expected: Any) -> None:
```

sugatoray replied:
Done!

sugatoray commented:
@Jasha10 I also added a section for clearing resolvers in the docs. Please take a look at it and let me know if you think changes should be made.

Note: *I am not sure if `OmegaConf.clear_resolvers()` has any documentation available or not. So, I added it here as any user trying to add a custom resolver, would probably find it handy here to know how to clear/remove resolver(s).*

For quick review, I am attaching a screenshot of the generated docs. I added the following section in `docs/source/custom_resolvers.rst`.

![image](https://user-images.githubusercontent.com/10201242/127732493-0b355818-2bd7-45fc-910f-0e1b42cb378b.png)

Jasha10 reviewed (changes requested):
> Note: _I am not sure if `OmegaConf.clear_resolvers()` has any documentation available or not. 

You're right, there was no documentation before.

> So, I added it here as any user trying to add a custom resolver, would probably find it handy here to know how to clear/remove resolver(s)._

Great, thanks!


A few more comments below on the tests:

sugatoray commented on omegaconf/omegaconf.py:
```diff
@@ -467,6 +467,23 @@ def clear_resolvers() -> None:
+        :param name: Name of the resolver.
+        :return: A bool (``True`` if resolver is removed, ``False`` if not found before removing).
+        """
+        if cls.has_resolver(name):  # pragma: nocover
```
> Is this `pragma: nocover` comment necessary?

You are right. It is not necessary. One of the tests already checks for the scenario with a *non-existent-resolver*. So, that if statement is covered already. I will remove it. Thank you. Good catch!

sugatoray replied:
Removed it!

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,66 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+    # - see functions in: tests/interpolation/test_custom_resolvers.py
+    # Using restore_resolvers is preferred over using
+    # "OmegaConf.clear_resolvers()" in the tests (where necessary).
+    assert expected["pre_addition"] == OmegaConf.has_resolver(name)
```
Done.

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,66 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+    if OmegaConf.has_resolver(name):
+        assert OmegaConf.clear_resolver(name)
+    else:
+        assert not OmegaConf.clear_resolver(name)
```
Done.

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,66 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+    # Example use-cases:
+    # - see functions in: tests/interpolation/test_custom_resolvers.py
+    # Using restore_resolvers is preferred over using
+    # "OmegaConf.clear_resolvers()" in the tests (where necessary).
```
Removed.

sugatoray commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,66 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+            # expected
+            dict(pre_addition=True, post_addition=True),
+            id="remove-default-resolver",
+        ),
```
Removed.

sugatoray commented:
@Jasha10 Thank you, for the detailed review. I have made the changes you suggested. Instead of adding `expected_result` as another parameter, I added that as a key to the parameter, `expected`: 

```python
expected["result"]
```

Jasha10 commented on docs/source/custom_resolvers.rst:
```diff
@@ -391,3 +391,68 @@ as interpolations), and they return a ``ListConfig`` that contains keys or value
+    >>> # check if resolver exists (after removal)
+    >>> OmegaConf.has_resolver("str.lower")
+    False
+
```
```suggestion
    >>> # default resolvers are not affected
    >>> OmegaConf.has_resolver("oc.env")
    >>> True

```

Jasha10 commented on tests/test_omegaconf.py:
```diff
@@ -518,3 +518,42 @@ def test_resolve(cfg: Any, expected: Any) -> None:
+) -> None:
+    if cfg_params:
+        OmegaConf.register_new_resolver(**cfg_params)
+        assert expected["post_addition"] == OmegaConf.has_resolver(name)
```
Also, could you please rename `cfg_params` -> `register_resolver_params` and `post_addition` -> `pre_clear`?

sugatoray replied:
Thank you. Done!

Jasha10 reviewed (changes requested):
Looking good! Just a few more comments.

sugatoray commented:
@Jasha10 I just pused the changes to the PR. Thank you, for the reviews. Please let me know if there is anything else that needs change.

omry commented on omegaconf/omegaconf.py:
```diff
@@ -467,6 +467,23 @@ def clear_resolvers() -> None:
+        :return: A bool (``True`` if resolver is removed, ``False`` if not found before removing).
+        """
+        if cls.has_resolver(name):
+            _ = BaseContainer._resolvers.pop(name)
```
Why not just ignore the return value?
```suggestion
            BaseContainer._resolvers.pop(name)
```

omry commented on omegaconf/omegaconf.py:
```diff
@@ -467,6 +467,23 @@ def clear_resolvers() -> None:
+        """
+        if cls.has_resolver(name):
+            _ = BaseContainer._resolvers.pop(name)
+            return not cls.has_resolver(name)
```
This should never be False. I think we can trust dict pop to work.
```suggestion
            return True
```

sugatoray replied:
Ok. I see your point. Incorporating your suggestion. Thank you.

> As a side note, the reason I kept that line `return not cls.has_resolver(name)` was to avoid any edge case, where `dict.pop()` fails somehow. However, I must also acknowledge that I am not aware of any such edge cases yet. So, I presume, a simple `return True` should suffice.

omry reviewed (commented):
Thanks for working on it. Overall looks good. I have a few comments.

sugatoray commented on docs/source/custom_resolvers.rst:
```diff
@@ -391,3 +391,71 @@ as interpolations), and they return a ``ListConfig`` that contains keys or value
+Clearing (removing) all custom resolvers at once is supported using ``OmegaConf.clear_resolvers()``.
+This method removes all custom resolvers, but retains all the default resolvers.
+
+.. code-block:: python
+
+    def clear_resolvers() -> None
+
+In the following example, first we register a new custom resolver ``str.lower``, and then clear all
+custom resolvers.
+
+.. doctest::
+
+    >>> # register a new resolver: str.lower
+    >>> OmegaConf.register_new_resolver(
+    ...     name='str.lower',
+    ...     resolver=lambda x: str(x).lower(),
+    ... )
+    >>> # check if resolver exists (after adding, before removal)
+    >>> OmegaConf.has_resolver("str.lower")
+    True
+    >>> # clear all custom-resolvers
+    >>> OmegaConf.clear_resolvers()
+    >>> # check if resolver exists (after removal)
+    >>> OmegaConf.has_resolver("str.lower")
+    False
+    >>> # default resolvers are not affected
+    >>> OmegaConf.has_resolver("oc.env")
+    True
```
Granted it's down right simple :wink: . But, I believe, examples are kind of a major useful feature of the documentation of this library. So, I thought if it is made clear through an example, it will be helpful. I wanted to show that custom resolvers are removed, while default ones are not, when `clear_resolvers()` is used.

However, if you insist on it's (the example: starting from line-412) removal, that's okay as well.

sugatoray commented:
@omry Thank you, for your reviews. I have made the changes you suggested (except the docs example on `clear_resolvers`). I hope, now it is very close to *ready-to-merge*.

omry reviewed (approved):
lgtm, thanks!
@Jasha10, feel free to merge.

sugatoray commented:
@omry and @Jasha10 It was a very pleasant experience working with you, on my first PR to this library. I hope to contribute more in future. 

💡 Question: is there any `CHANGELOG` that you would like me to update?

Jasha10 commented:
Thanks @sugatoray, it was good working with you!

Yes, good idea about the changelog (I forgot about that step).
I've created a follow-up PR #774 which adds a new file `news/769.feature`.

FYI: OmegaConf uses a tool called `towncrier` to generate its `NEWS.md` changelog. This news file gets updated during OmegaConf's release cycle; the contents of this `news/769.feature` will be added to `NEWS.md` automatically. The name of the news file `news/769.feature` refers to the original issue #769 that you created.
